sdk/storage/Azure.Storage.Internal.Avro/src/AvroParser.cs

```diff
@@ -0,0 +1,388 @@
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+#pragma warning disable SA1402 // File may only contain a single type
+
+namespace Azure.Storage.Internal.Avro
+{
+    internal static class AvroParser
+    {
+        public static async Task<byte[]> ReadFixedBytesAsync(
+            Stream stream,
+            int length,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte[] data = new byte[length];
+            int start = 0;
+            while (length > 0)
+            {
+                int n = async ?
+                    await stream.ReadAsync(data, start, length, cancellationToken).ConfigureAwait(false) :
+                    stream.Read(data, start, length);
+                start += n;
+                length -= n;
+
+                // We hit the end of the stream
+                if (n <= 0)
+                    return data;
+            }
+            return data;
+        }
+
+        private static async Task<byte> ReadByteAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte[] bytes = await ReadFixedBytesAsync(stream, 1, async, cancellationToken).ConfigureAwait(false);
+            return bytes[0];
+        }
+
+        // Stolen because the linked references in the Avro spec were subpar...
+        private static async Task<long> ReadZigZagLongAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte b = await ReadByteAsync(stream, async, cancellationToken).ConfigureAwait(false);
+            ulong next = b & 0x7FUL;
+            int shift = 7;
+            while ((b & 0x80) != 0)
+            {
+                b = await ReadByteAsync(stream, async, cancellationToken).ConfigureAwait(false);
+                next |= (b & 0x7FUL) << shift;
+                shift += 7;
+            }
+            long value = (long)next;
+            return (-(value & 0x01L)) ^ ((value >> 1) & 0x7fffffffffffffffL);
+        }
+
+        public static Task<object> ReadNullAsync() => Task.FromResult<object>(null);
+
+        public static async Task<bool> ReadBoolAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte b = await ReadByteAsync(stream, async, cancellationToken).ConfigureAwait(false);
+            return b != 0;
+        }
+
+        public static async Task<long> ReadLongAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken) =>
+            await ReadZigZagLongAsync(stream, async, cancellationToken).ConfigureAwait(false);
+
+        public static async Task<int> ReadIntAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken) =>
+            (int)(await ReadLongAsync(stream, async, cancellationToken).ConfigureAwait(false));
```

Comment:
Do we want to do any checking here (make sure the num is int sized) or can we just let the cast fail?

Reply:
I think its fine to let the cast fail.